spec/uvarint_spec.cr

```diff
@@ -1,77 +1,191 @@
+      v = UVarInt.new r
+      u = v.uint
+      u.should eq(0)
+    end
+
+    it "handles strings" do
+      s = "abcd"
+      v = UVarInt.new s
+      u = v.uint
+      u.should eq(97)
     end
-  end
 
-  it "decodes single bytes" do
-    n = Random.rand(0..0x7F).to_u8
-    buf = [n]
-    d = UVarint.decode buf
-    d.should eq(n)
+    it "raises an exception if passed more than 10 bytes" do
+      b = Array.new(20, 0_u8)
+      ex = expect_raises(ArgumentError) { UVarInt.new b }
+      ex.message.should eq("cannot initialize with more than 10 bytes")
+    end
   end
 
-  it "decodes multiple bytes" do
-    buf = [0xAC_u8, 0x02_u8]
-    d = UVarint.decode buf
-    d.should eq(300)
+  describe "[] macro" do
+    it "throws if more than 10 bytes are passed" do
+      ex = expect_raises(ArgumentError) { UVarInt[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0] }
+      ex.message.should eq("cannot initialize with more than 10 bytes")
+    end
+
+    it "instantiates a UVarInt with the passed bytes" do
+      v = UVarInt[172, 2]
+      b = v.bytes
+      b.should eq(Bytes[172, 2])
+    end
   end
 
-  it "decodes multiple bytes with zero" do
-    n = Random.rand(0x7F).to_u8
-    buf = [0x80_u8, n]
-    d = UVarint.decode buf
-    d.should eq(n.to_u64 << 7)
+  describe "decode" do
+    it "decodes single bytes" do
+      n = Random.rand(0..0x7F).to_u8
+      b = Bytes[n]
+      v = UVarInt.new b
+      u = v.uint
+      u.should eq(n)
+    end
+
+    it "decodes multiple bytes" do
+      b = Bytes[0xAC_u8, 0x02_u8]
+      v = UVarInt.new b
+      u = v.uint
+      u.should eq(300)
+    end
+
+    it "decodes multiple bytes with zero" do
+      n = Random.rand(0x7F).to_u8
+      b = Bytes[0x80_u8, n]
+      v = UVarInt.new b
+      u = v.uint
+      u.should eq(n.to_u64 << 7)
+    end
   end
 
-  it "encodes to single bytes" do
-    n = Random.rand(0x7F).to_u64
-    e = UVarint.encode n
-    e.should eq([n.to_u8])
+  describe "encode" do
+    it "encodes to single bytes" do
+      n = Random.rand(0x7F).to_u64
+      v = UVarInt.new n
+      b = v.bytes
+      b.should eq(Bytes[n.to_u8])
+    end
+
+    it "encodes to multiple bytes" do
+      v = UVarInt.new 300_u64
+      b = v.bytes
+      b.should eq(Bytes[0xAC_u8, 0x02_u8])
+    end
+
+    it "encodes to multiple bytes with first byte zero" do
+      v = UVarInt.new 0x0F00_u64
+      b = v.bytes
+      b.should eq(Bytes[0x80_u8, 0x1E_u8])
+    end
   end
 
-  it "encodes to multiple bytes" do
-    e = UVarint.encode 300_u64
-    e.should eq([0xAC_u8, 0x02_u8])
+  describe "to_s" do
+    it "returns self.uint.to_s" do
+      v = UVarInt.new 300_u64
+      s = v.to_s 16
+      s.should eq("12c")
+    end
   end
 
-  it "encodes to multiple bytes with first byte zero" do
-    n = 0x0F00_u64
-    e = UVarint.encode n
-    e.should eq([0x80_u8, 0x1E_u8])
+  describe "hexstring" do
+    it "returns self.bytes.hexstring" do
+      v = UVarInt.new 300_u64
+      h = v.hexstring
+      h.should eq("ac02")
```

Comment:
`Bytes.hexstring` returns the hex representation of the bytes, whereas `to_s` returns the hex representation of the uint.